docs: add a Customize your template series under Get started (#26712)
coder/coder: docs/tutorials/quickstart/personalize-with-dotfiles.md

## What you'll do

- ✅ Add the Coder dotfiles module to the template.
- ✅ Point a workspace at your dotfiles repository.
- ✅ Apply your shell and editor settings when a workspace starts.

## Modules in brief

A [module](https://developer.hashicorp.com/terraform/language/modules) is a reusable bundle of Terraform you pull in by reference instead of writing it yourself.
The [Coder Registry](https://registry.coder.com/modules) publishes modules for common workspace features,
and the [dotfiles module](https://registry.coder.com/modules/coder/dotfiles) is one of them.

A `module` block names a `source` to pull from and a `version` to pin:

```tf
# --- Dotfiles ---

module "dotfiles" {
  count    = data.coder_workspace.me.start_count
  source   = "registry.coder.com/coder/dotfiles/coder"
  version  = "1.4.2"
  agent_id = coder_agent.main.id
}
```

`count = data.coder_workspace.me.start_count` builds the module only when the workspace is running,
and `agent_id` attaches it to the workspace agent so the module runs inside the workspace.

## Step 1: Add the dotfiles module

> Add the `module` block above anywhere in `main.tf`,

Author: It's not clear the way this is worded that users were supposed to add the Terraform block from lines 31 to 38 in main.tf. I suggest pulling lines 28 to 42 into step 1 and reorganizing it so it's more clear that we're asking users to put that module block in the file.

Author: From Coder Agents: Done in dbc46eecee. The `module` block and its `count`/`agent_id` explanation now live inside Step 1 under a clear "Add this `module` block anywhere in `main.tf`" instruction, so it no longer reads as code shown for its own sake. "Modules in brief" keeps only the conceptual definition.

then publish a new version of the template:

<div class="tabs">

### UI

In the web editor, add the `module` block to `main.tf`.
Select **Build**, wait for the build to pass, then select **Publish**.

### CLI

Add the `module` block to `~/coder-quickstart/main.tf`,
then publish a new version:

```sh
coder templates push -d ~/coder-quickstart -y quickstart
```

</div>

> Create a workspace from the template.

Author: While it's good to cater to folks who are just starting fresh on this page and skipped the previous page, I want this to be geared more toward the people who are working with the same template and the same workspace. Here's the flow:

- User publishes the new template version.
- User then updates and restarts the workspace.
- User then goes in to modify the parameters for the workspace and notices the two additional parameters for dotfiles.
- User fills those out and restarts the workspace to have their dotfiles apply.

Author: From Coder Agents: Done in dbc46eecee. New Step 2, "Apply your dotfiles to your workspace," follows the flow you described: publish the new version in Step 1, then update the existing workspace, set the two new dotfiles parameters, and restart to apply. It mirrors the update pattern from Add a programming language (UI update form, or `coder update <your-workspace> --always-prompt`).

The creation form now shows 2 new fields:

- **Dotfiles URL**, the repository to clone. The module accepts SSH (`git@host:user/repo`) and HTTPS (`https://host/user/repo`) URLs.
- **Dotfiles Branch**, the branch to apply.

Enter a dotfiles repository URL,
create the workspace,
and the module clones the repository and applies it at startup.

## Step 2: Configuration first, tools optional

Dotfiles are mainly for configuration that lives in your home directory,

> the part of the workspace that [persists across rebuilds](./install-command-line-tools.md#what-persists-in-a-workspace).

Author: ```suggestion
Dotfiles are mainly for configuration that lives in your home directory, the part of the workspace that [persists across rebuilds](./install-command-line-tools.md#what-persists-in-a-workspace).
```

We're doing one sentence per line, not sembr.

Author: From Coder Agents: Done in dbc46eecee. Applied your suggestion verbatim and reflowed the rewritten guide to one sentence per line (no sembr).

# Personalize your workspace with dotfiles

This guide is for a developer who finished [Launch your first workspace](./launch-workspace.md)
and wants every workspace to start with their own shell, editor, and tool configuration.

[Dotfiles](https://dotfiles.github.io/) are the hidden configuration files in your home directory,
such as `.bashrc`, `.gitconfig`, and `.vimrc`.

> In this guide you add the Coder dotfiles module to the template,

Author: ```suggestion
In this guide, you add the Coder dotfiles module to the template,
```

Author: From Coder Agents: Done in dd5d3ab6ad. Added the comma. The surrounding sentence is reworked to drop the Homebrew framing, but it keeps your "In this guide, you add the Coder dotfiles module to the template" phrasing.

watch a common bootstrap mistake fail,
and keep the personalization that survives a workspace rebuild.

> [!NOTE]
> This guide assumes your Quickstart template is open for editing.
> If it is not, refer to [Customize workspace startup](./customize-workspace-startup.md#open-the-template-for-editing).

## What you'll do

- ✅ Add the Coder dotfiles module to the template.

> - ✅ See why a dotfiles script that installs Homebrew packages fails.

Author: ```suggestion
- ✅ Learn why a dotfiles script that installs Homebrew packages fails.
```

Author: From Coder Agents: This bullet is gone in dd5d3ab6ad. The Homebrew failure moved out of the dotfiles guide entirely (see the next thread), so the "why ... Homebrew ... fails" item went with it. The dotfiles "What you'll do" list is now config-focused.

and the [dotfiles module](https://registry.coder.com/modules/coder/dotfiles) is one of them.

A `module` block names a `source` to pull from and a `version` to pin:

```tf
# --- Dotfiles ---

module "dotfiles" {
  count    = data.coder_workspace.me.start_count
  source   = "registry.coder.com/coder/dotfiles/coder"
  version  = "1.4.2"
  agent_id = coder_agent.main.id
}
```

`count = data.coder_workspace.me.start_count` builds the module only when the workspace is running,
and `agent_id` attaches it to the workspace agent so the module runs inside the workspace.

## Step 1: Add the dotfiles module

Add the `module` block above anywhere in `main.tf`,
then push a new version of the template:

```console
coder templates push -d ~/coder-quickstart -y quickstart
```

Create a workspace from the template.
The creation form now shows 2 new fields:

- **Dotfiles URL**, the repository to clone. The module accepts SSH (`git@host:user/repo`) and HTTPS (`https://host/user/repo`) URLs.
- **Dotfiles Branch**, the branch to apply.

Enter a dotfiles repository URL,
create the workspace,
and the module clones the repository and applies it at startup.

> ## Step 2: Watch a Homebrew bootstrap fail

Author: I know I suggested adding failure scenarios, but this one seems heavy-handed and out of scope of the whole document. I suggest moving this to the Homebrew/Mise document we'll eventually write. If we perhaps need to change the order so the Homebrew/Mise document comes before this document, that's fine as well.

Author: From Coder Agents: Done in dd5d3ab6ad. Moved the Homebrew failure into a new guide, "Install your own command-line tools," and sequenced it before the dotfiles guide, exactly as you suggested. That guide shows `brew install` failing, then mise installing the tool into `/home/coder` so it survives a rebuild. I tested the whole flow against `codercom/enterprise-base:ubuntu`, including a container-rebuild persistence check on the `/home/coder` volume. The dotfiles guide is now config-only and cross-links the new guide for tools.

```text
install.sh: line 1: brew: command not found
```

This failure has 2 causes:

- **The image has no Homebrew**: the Quickstart workspace runs the `codercom/enterprise-base:ubuntu` image, which ships `apt-get`, `git`, and `python3`, but not Homebrew. The dotfiles script cannot find `brew`.
- **Persistence**: even if you installed Homebrew and a tool at startup, both land outside your home directory. The Quickstart workspace keeps `/home/coder` on a persistent volume, but rebuilds the rest of the container from the image on every start, so a tool installed elsewhere disappears on the next rebuild.

Following the laptop pattern steered the workspace wrong.

## Step 3: Personalize what persists

Keep dotfiles focused on configuration that lives in your home directory,
which is the part of the workspace that persists.
Symlinking `.bashrc`, `.gitconfig`, and editor settings is exactly what the dotfiles module is for.

For a tool you need in every workspace,
install it where it persists across rebuilds:

- Add it to the template's startup script with `apt-get`, as in [Add a programming language](./add-a-language.md). This installs the tool for everyone who uses the template.

> - Or bake it into the workspace image, so it is present before startup runs.

Author: ```suggestion
install it where it persists across rebuilds.
You have a few options to make it persist across rebuilds:

- Add it to the template's startup script with `apt-get`, as in [Add a programming language](./add-a-language.md). This installs the tool for everyone who uses the template.
- Bake it into the workspace image, so it is present before startup runs.
```

Author: From Coder Agents: Done in dd5d3ab6ad. Reworked into a lead-in plus options with no leading "Or." In the dotfiles guide the two options now point to mise (for a personal tool) and the template startup script (for a team-wide tool); the new mise guide carries the same "startup script or bake it into the image" guidance for the everyone-needs-it case.

Remove the `brew install` line from `install.sh`,
keep your configuration symlinks,
and create the workspace again.
The dotfiles apply cleanly,

> and your shell and editor settings are there from the first prompt.

Author: Given I don't want to reference Homebrew in this doc, rework this sentence.

Author: From Coder Agents: Done in dd5d3ab6ad. Homebrew is no longer referenced anywhere in the dotfiles guide; this sentence is reworked around keeping dotfiles to configuration.

## What just happened

You pulled a whole feature into the template with a few lines by referencing a module.
You also found the limit of startup-time personalization:

- A workspace startup script cannot rely on tools, like Homebrew, that the workspace image does not provide.

> - Only your home directory survives a rebuild, so durable tool installs belong in the template, not in dotfiles.

Author: Rework this.

Author: From Coder Agents: Done in dd5d3ab6ad. Reworked as part of making the dotfiles guide config-only; the "What just happened" recap no longer mentions Homebrew or startup-time tool installs.

# Personalize your workspace with dotfiles

> This guide is for a developer who finished [Launch your first workspace](./launch-workspace.md)

Author: Same nit

Author: From Coder Agents: Direct-address intro here too, in 39b5f58e76.

and wants every workspace to start with their own shell, editor, and Git configuration.

[Dotfiles](https://dotfiles.github.io/) are the hidden configuration files in your home directory,
such as `.bashrc`, `.gitconfig`, and `.vimrc`.
In this guide, you add the Coder dotfiles module to the template,
point a workspace at your dotfiles repository,
and apply your shell and editor settings every time a workspace starts.

> [!NOTE]
> This guide assumes your Quickstart template is open for editing.

> > If it is not, refer to [Customize workspace startup](./customize-workspace-startup.md#open-the-template-for-editing).

Author: Same nit about contractions

Author: From Coder Agents: Contracted ("If it's not") in 39b5f58e76, consistent with the other pages.

Add the `module` block above anywhere in `main.tf`,
then push a new version of the template:

```sh
coder templates push -d ~/coder-quickstart -y quickstart
```

Create a workspace from the template.
The creation form now shows 2 new fields:

- **Dotfiles URL**, the repository to clone. The module accepts SSH (`git@host:user/repo`) and HTTPS (`https://host/user/repo`) URLs.
- **Dotfiles Branch**, the branch to apply.

Enter a dotfiles repository URL,
create the workspace,
and the module clones the repository and applies it at startup.

## Step 2: Keep dotfiles focused on configuration

Dotfiles are for configuration that lives in your home directory,
the part of the workspace that [persists across rebuilds](./install-tools-with-mise.md#what-persists-in-a-workspace).
Symlinking `.bashrc`, `.gitconfig`, and editor settings is exactly what the dotfiles module is for.

A dotfiles repository can also run an `install.sh` script,
which tempts you to install tools from it.
Keep tools out of dotfiles,
and install each tool where it persists instead:

- For a tool that is yours alone, use [mise](./install-tools-with-mise.md), which installs into your home directory.

> - For a tool everyone on the template needs, add it to the startup script with `apt-get`, as in [Add a programming language](./add-a-language.md).

Author: I don't necessarily agree with the sentiment that dotfiles aren't meant for tools. I use Chezmoi for my own personal dotfiles, and I have a Brewfile configured to store the formulae I want to use in my own workspace.

This doc should be equipped, based on the installing custom tools doc, to handle tool installs from a dotfiles repo.

Author: From Coder Agents: Agreed, and reworked in 39b5f58e76. Step 2 is now "Configuration first, tools optional": dotfiles can install tools (for example a Brewfile or a mise config via install.sh), as long as each tool lands somewhere that persists. It cross-links Install your own command-line tools for the persistence rules instead of forbidding the pattern.
